src/Proto3/Suite/DotProto/Generate.hs

```diff
@@ -1539,115 +1187,114 @@ dotProtoEnumD
-              ]
-
-         toJSONPBDecl =
-           match_ (HsIdent "toJSONPB") [ patVar "x", HsPWildCard ]
-             (HsUnGuardedRhs
-                (HsApp (HsVar (jsonpbName "enumFieldString"))
-                       (HsVar (unqual_ "x"))))
-             []
-
-         toEncodingPBDecl =
-           match_ (HsIdent "toEncodingPB") [ patVar "x", HsPWildCard ]
-             (HsUnGuardedRhs
-                (HsApp (HsVar (jsonpbName "enumFieldEncoding"))
-                       (HsVar (unqual_ "x"))))
-             []
-
-     pure [ dataDecl_ enumName
-                      [ conDecl_ (HsIdent con) [] | con <- enumConNames ]
-                      defaultEnumDeriving
-          , namedInstD enumName
-          , instDecl_ (haskellName "Enum") [ type_ enumName ]
-                      [ HsFunBind toEnumD, HsFunBind fromEnumD
-                      , HsFunBind succD, HsFunBind predD ]
-          , instDecl_ (jsonpbName "ToJSONPB") [ type_ enumName ]
-                      [ HsFunBind [toJSONPBDecl]
-                      , HsFunBind [toEncodingPBDecl]
-                      ]
-          , instDecl_ (jsonpbName "FromJSONPB") [ type_ enumName ]
-                      [ HsFunBind parseJSONPBDecls ]
-          -- Generate Aeson instances in terms of JSONPB instances
-          , toJSONInstDecl enumName
-          , fromJSONInstDecl enumName
+      toEnumD = toEnumDPatterns <> [ toEnumFailure ]
+      fromEnumD =
+          [ match_ (HsIdent "fromEnum") [ HsPApp (unqual_ conName) [] ]
+                   (HsUnGuardedRhs (intE conIdx))
+                   []
+          | (conIdx, conName) <- enumCons
+          ]
+      succD = zipWith succDPattern enumConNames (tail enumConNames) <> [ succFailure ]
+      predD = zipWith predDPattern (tail enumConNames) enumConNames <> [ predFailure ]
+
+      toEnumDPatterns =
+          [ match_ (HsIdent "toEnum")
+                   [ intP conIdx ]
+                   (HsUnGuardedRhs (uvar_ conName)) []
+          | (conIdx, conName) <- enumCons ]
+
+      succDPattern thisCon nextCon =
+          match_ (HsIdent "succ") [ HsPApp (unqual_ thisCon) [] ]
+                 (HsUnGuardedRhs (uvar_ nextCon)) []
+      predDPattern thisCon prevCon =
+          match_ (HsIdent "pred") [ HsPApp (unqual_ thisCon) [] ]
+                 (HsUnGuardedRhs (uvar_ prevCon)) []
+
+      toEnumFailure   = match_ (HsIdent "toEnum") [ HsPVar (HsIdent "i") ]
+                               (HsUnGuardedRhs
+                                  (apply toEnumErrorE [enumNameE , uvar_ "i" , boundsE]))
+                               []
+      succFailure     = match_ (HsIdent "succ") [ HsPWildCard ]
+                               (HsUnGuardedRhs (HsApp succErrorE enumNameE)) []
+      predFailure     = match_ (HsIdent "pred") [ HsPWildCard ]
+                               (HsUnGuardedRhs (HsApp predErrorE enumNameE)) []
+
+      parseJSONPBDecls :: [HsMatch]
+      parseJSONPBDecls = foldr ((:) . matchConName) [mismatch] enumConNames
```

Comment:
Maybe `map matchConName enumConNames <> [ mismatch ]`

Reply:
the `foldr` is more efficient, though it doesn't matter here...